common/flatpak-dir.c

```diff
@@ -7652,31 +7694,87 @@ flatpak_dir_install (FlatpakDir          *self,
         }
       else
         {
+          /* If the system-helper can find bindfs on the system, it should return a child repo for us
+             to pull into. The child repo resides inside the system's repo/tmp/flatpak-cache-$ref-XXXXXX
+             directory. This directory is mounted using bindfs which enables us to  pull into it. Once the pull
+             completes, the directory is unmounted before "Deploy" to avoid modifications and do a trusted pull
+             from this child repo to the system repo via hardlinking.
+
+             If this operation is unsupported or failed, the pull is then tried with the fallback codepath. */
+
+          g_autofree gchar *helper_child_repo_path = NULL;
+          g_autoptr(OstreeRepo) child_repo = NULL;
+          g_auto(GLnxLockFile) child_repo_lock = { 0, };
+          g_autoptr(GError) local_error = NULL;
+          gboolean fallback = FALSE;
+
+          if (!flatpak_dir_system_helper_get_child_repo_for_pull (self, ref, opt_commit,
+                                                                  installation ? installation : "",
+                                                                  &helper_child_repo_path, cancellable, &local_error))
+            {
+              if (g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_UNKNOWN_METHOD) ||
+                  g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED))
+                g_debug ("Failed to get child repo in repo/tmp: %s", local_error->message);
+              else
+                g_warning ("Failed to get child repo in repo/tmp: %s", local_error->message);
+            }
+          else
+            {
+              g_autoptr(GFile) repo_file = g_file_new_for_path (helper_child_repo_path);
+              g_autoptr(OstreeRepo) repo = ostree_repo_new (repo_file);
+              if (!ostree_repo_open (repo, NULL, &local_error))
+                {
+                  g_warning ("Failed to open child repo at %s: %s", helper_child_repo_path, local_error->message);
+
+                  /* Explicitly clearing the child_repo pointer closes any open FDs on the child_repo.
+                     This prevents EBUSY error failure during unmounting. */
+                  g_clear_object (&repo);
+                  flatpak_dir_cancel_pull (self, cancellable, helper_child_repo_path);
+                }
+              else
+                child_repo = g_steal_pointer (&repo);
+            }
+
           /* We're pulling from a remote source, we do the network mirroring pull as a
              user and hand back the resulting data to the system-helper, that trusts us
              due to the GPG signatures in the repo */
 
-          child_repo = flatpak_dir_create_system_child_repo (self, &child_repo_lock, NULL, error);
+          /* Fallback */
```

Comment:
This comment would be more useful if it stated under what conditions we fall back. E.g:

The system helper failed to make a repo available to us (probably because bindfs wasn't found). Fall back to doing things the old way.